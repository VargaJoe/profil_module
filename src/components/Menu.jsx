import React, { Component } from 'react';
import MenuItem from './MenuItem';

class Menu extends Component {
	constructor(props) {
		super(props);
		this.state = {
			menuName : ["Item 1", "Item 2", "Item 3", "Item 4", "Item 5", "Item 6", "Item 7", "Item 8"],
			menuIcon : ["fi flaticon-lightbulb-idea", "fi flaticon-family-tree", "fi flaticon-partners", "fi flaticon-calendar", "fi flaticon-discussion", "fi flaticon-people", "fi flaticon-folded-newspaper", "fi flaticon-book"],
		};
	}

	render () {
		return (
			<div className="sidebar__menu">
				<MenuItem name={this.state.menuName[0]} icon={this.state.menuIcon[0]} />
				<MenuItem name={this.state.menuName[1]} icon={this.state.menuIcon[1]} />
				<MenuItem name={this.state.menuName[2]} icon={this.state.menuIcon[2]} />
				<MenuItem name={this.state.menuName[3]} icon={this.state.menuIcon[3]} />
				<MenuItem name={this.state.menuName[4]} icon={this.state.menuIcon[4]} />
				<MenuItem name={this.state.menuName[5]} icon={this.state.menuIcon[5]} />
				<MenuItem name={this.state.menuName[6]} icon={this.state.menuIcon[6]} />
				<MenuItem name={this.state.menuName[7]} icon={this.state.menuIcon[7]} />
			</div>
		)
	}
};

export default Menu;
