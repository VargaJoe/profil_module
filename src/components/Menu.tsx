import * as React				            from 'react';
import MenuItem 				            from './MenuItem';
import { connect }                          from 'react-redux';
import { Actions
}                                           from '@sensenet/redux';
import { LoginState }                       from '@sensenet/client-core';
import { IODataParams }                     from '@sensenet/client-core/dist/Models/IODataParams';
import { MenuItemType }                     from '../type/MenuItemType';
const DATA = require('../config.json');
const fontImportantClass = ' fi ';

interface Props {
    userLoginState: string;
    logoutEvent: Function;
    getMenuItems: (path: string, options: IODataParams<MenuItemType>) => Promise<{
        value: {
            entities: any;
            result: any;
        }
    }>;
}

export class MenuComponent extends React.Component<Props, any> {
    unmount: boolean;
	constructor(props: any) {
		super(props);
		this.state = {
			menuItems: null,
            isDataFetched: false,
            status: this.props.userLoginState !== LoginState.Authenticated
        };
        
        this.handleLogoutClick = this.handleLogoutClick.bind(this);
    }
    
    componentWillUnmount() {
        this.unmount = true;
    }

	public async componentDidMount  () {
        let path = DATA.menu;
        let menuItems = await this.props.getMenuItems(path, {
            select : ['Name', 'Id', 'Path', 'DisplayName', 'IconClassName']
        });
        if (!this.unmount) {
            this.setState({
                isDataFetched : true,
                menuItems: menuItems.value.entities.entities
            });  
        }
    }

    handleLogoutClick = (e: any) => {
        console.log(12);
    }

	public render () {
		if ( !this.state.isDataFetched ) {
            return null;
		}
        let menuItems = this.state.menuItems;
        const menu = Object.keys(menuItems).map( (key: any) => 
            (
                <MenuItem key={key} name={menuItems[key].DisplayName} icon={fontImportantClass + this.state.menuItems[key].IconClassName} pathTo="/"/>
            )
        );
		return (
			<div className="sn_sidebar__menu">
				{this.state.status ? '' : 
                    <MenuItem name={'All users'} icon={fontImportantClass + 'flaticon-group-of-businessmen'} pathTo="/otherUser"/>
                }
                {this.state.status ? '' : 
                    <MenuItem name={'Change Password'} icon={fontImportantClass + 'flaticon-group-of-businessmen'} pathTo="/changePass"/>
                }
                {this.state.status ? '' : menu}
                {this.state.status ? 
                    <MenuItem name={'Login'} icon={fontImportantClass + 'flaticon-folded-newspaper'} pathTo="/login"/>
                    : 
                    <MenuItem name={'Log Out'} logoutEvent={this.handleLogoutClick} icon={fontImportantClass + 'flaticon-group-of-businessmen'} pathTo="/"/>
                }
				
			</div>
		);
	}
}

// export default Menu;

const mapStateToProps = (state: any, match: any) => {
    return {
        userLoginState: 		state.sensenet.session.loginState, // state.user.user.FullName,
    };
};

export default connect(
    mapStateToProps,
    (dispatch) => ({
        getMenuItems:       (path: string, options: any) => dispatch(Actions.requestContent( path, options )),
        logout:             () => dispatch(Actions.userLogout()),
    })
)(MenuComponent as any);