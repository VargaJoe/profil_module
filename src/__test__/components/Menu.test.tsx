jest.unmock('../../index.tsx');
jest.unmock('redux-mock-store');

import * as React 						        from 'react';
import Menu 							        from '../../components/Menu';
import { 
	configure, 
    shallow,
    mount,
    render, 
    ReactWrapper, 
    ShallowWrapper } 							from 'enzyme';
import * as Adapter 					        from 'enzyme-adapter-react-16';
import * as createMockStore 				    from 'redux-mock-store';
import toJson 				      		        from 'enzyme-to-json';
import { Actions
}                                               from '@sensenet/redux';
import { LoginState }                           from '@sensenet/client-core';
import { Provider } 							from 'react-redux';
import { Repository }                           from '@sensenet/client-core';
import { JwtService }                           from '@sensenet/authentication-jwt';
import { promiseMiddleware }                    from '@sensenet/redux-promise-middleware';

const DATA = require('../../config.json');
let path = DATA.domain + DATA.menu;
const  options = {
    select : ['Name', 'IconName', 'Id', 'Path', 'DisplayName']
};
configure( {adapter: new Adapter()} );
describe('<Menu /> shallow rendering with user data', () => {
    let store, menu: ReactWrapper<any, any>;
    let _store;
    let repo;
    let _jwtService;
    const contentMockResponse = {
        ok: true,
        status: 200,
        json: async () => {
            return {
                d: {
                    entities: {},
                    result: {
                        3892: {
                          'Id': 3892,
                          'Path': '/Root/Sites/Profil/Menu/MenuItem1',
                          'Name': 'MenuItem1',
                          'Type': 'ProfilMenuItem',
                          'IconName': 'flaticon-pirate',
                          'DisplayName': 'Menu Item 1'
                        },
                        3891: {
                          'Id': 3891,
                          'Path': '/Root/Sites/Profil/Menu/MenuItem100',
                          'Name': 'MenuItem100',
                          'Type': 'ProfilMenuItem',
                          'IconName': 'flaticon-island',
                          'DisplayName': 'Menu Item 100'
                        },
                        3893: {
                          'Id': 3893,
                          'Path': '/Root/Sites/Profil/Menu/Menu Item 2',
                          'Name': 'Menu Item 2',
                          'Type': 'ProfilMenuItem',
                          'IconName': 'flaticon-youtube',
                          'DisplayName': 'Menu Item 2'
                        }
                    },
                },
            };
        },
    } as Response;
    
	beforeEach( () => {
        repo = new Repository({ repositoryUrl: DATA.domain}, async () => contentMockResponse);
        _jwtService = new JwtService(repo);
        const mockStore = (createMockStore as any)([promiseMiddleware(repo)]);
        store = mockStore({
			sensenet: {
				session: {
					user: {
						userName: 'TestUser1'
                    },
                    loginState: 'Authorized'
				}
			}
		});
        menu = mount(
            <Provider store={store}>
                <Menu /> 
            </Provider>
        );   
    }); 
    // test Snapshot 
	it('Match to snapshot', () => {
        expect(toJson(menu)).toMatchSnapshot();
    });

    let data;
    let dataWithoutOptions;
    
    beforeEach(async (done) => {
        let menuContent = menu.children().children();
        // data = await menuContent.props().getMenuItems(DATA.menu, options);
        // console.log(menuContent.props().getMenuItems(DATA.menu, options));
    });
    // it('should return mockdata', () => {
    //     let menuContent = menu.children().children();
    //     console.log(menuContent.props().getMenuItems(DATA.menu, options));
    //     expect(data).toBe({ entities: {}, result: [] });
    // });

    // it('Menu have sidebar class', () => {
    //     let menuContent = menu.children().children();
    //     expect(menuContent.find('.sn_sidebar__menu').length).toBe(0);
    // });

    // it('call ComponentDidMouint', () => {
    //     jest.spyOn(Menu.prototype, 'componentDidMount');
    //     expect(Menu.prototype.componentDidMount.call.length).toBe(1);
    // });

    // it('Props state value should be the same as in store', () => {
    //     let menuContent = menu.children().children();
    //     expect(menuContent.props().userLoginState).toBe('Authorized');
	// });
    
});
 