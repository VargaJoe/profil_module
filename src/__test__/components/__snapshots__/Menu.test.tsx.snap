// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> rendering Match to snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <MenuComponent
      getMenuItems={[Function]}
      userLoginState="alma"
    />
  </Router>
</BrowserRouter>
`;

exports[`<Menu /> rendering Match to snapshot with data 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <MenuComponent
      getMenuItems={[Function]}
      userLoginState="alma"
    >
      <div
        className="sn_sidebar__menu"
      >
        <MenuItem
          icon=" fi flaticon-folded-newspaper"
          name="Login"
          pathTo="/login"
        >
          <Link
            className="sn_sidebar__menu__item"
            onClick={[Function]}
            replace={false}
            to="/login"
          >
            <a
              className="sn_sidebar__menu__item"
              href="/login"
              onClick={[Function]}
            >
              <span
                className="sn_sidebar__menu__item__icon"
              >
                <i
                  className=" fi flaticon-folded-newspaper"
                />
              </span>
              <span
                className="sn_sidebar__menu__item__name"
              >
                Login
              </span>
            </a>
          </Link>
        </MenuItem>
      </div>
    </MenuComponent>
  </Router>
</BrowserRouter>
`;

exports[`<Menu /> rendering Match to snapshot with status = true  1`] = `
<div
  className="sn_sidebar__menu"
>
  <MenuItem
    icon=" fi flaticon-folded-newspaper"
    name="Login"
    pathTo="/login"
  />
</div>
`;
